### ISSUE TITLE
`IGC_ForceOCLSIMDWidth=16` Breaks ALL Masked Triton Stores on BMG

### ISSUE THREAD

**Issue by ofirzaf** [intel/intel-xpu-backend-for-triton#6206]
## Summary

Setting `IGC_ForceOCLSIMDWidth=16` causes the Intel Graphics Compiler (IGC) to
miscompile **every** Triton kernel that uses masked `tl.store()` when `BLOCK_SIZE > 16`
on Battlemage (BMG) GPUs. Every other 16-element SIMD sub-group silently produces
**zero output** — the stores are never executed. This is not kernel-specific; even the
simplest possible 3-line kernel is affected.

We found this issue while trying to debug speculative decoding in vLLM. The issue is not affecting vLLM anymore via vllm-project/vllm#35298

## Minimal Reproducer

```python
import triton
import triton.language as tl
import torch

@triton.jit
def trivial_masked_store(out_ptr, N: tl.constexpr):
    j = tl.arange(0, N)
    mask = j < 59
    tl.store(out_ptr + j, j, mask=mask)

device = torch.device("xpu")
out = torch.full((64,), -999, dtype=torch.int32, device=device)
trivial_masked_store[(1,)](out, N=64)
torch.xpu.synchronize()
print(out.cpu().tolist())
```

### Expected output
```
[0, 1, 2, ..., 57, 58, -999, -999, -999, -999, -999]
```

### Actual output with `IGC_ForceOCLSIMDWidth=16`

```
[0, 1, 2, ..., 14, 15, -999, -999, ..., -999, 32, 33, ..., 46, 47, -999, ..., -999]
```

Sub-groups 0 (`[0:16]`) and 2 (`[32:48]`) are correct.
Sub-groups 1 (`[16:32]`) and 3 (`[48:64]`) retain their initial values — stores were
never executed.

Here is a script to help reproduce the issue [repro_minimal.py](https://github.com/user-attachments/files/25573986/repro_minimal.py)

```bash
python repro_minimal.py  # no issue
IGC_ForceOCLSIMDWidth=16 python repro_minimal.py  # issue
```

## Failure Pattern

The corruption follows a strict 16-element SIMD sub-group pattern:

| Sub-group index | Element range | Status |
|---|---|---|
| 0 (even) | `[0:16]` | **Correct** |
| 1 (odd)  | `[16:32]` | **Stores dropped** (output = initial value) |
| 2 (even) | `[32:48]` | **Correct** |
| 3 (odd)  | `[48:64]` | **Stores dropped** (output = initial value) |

### Environment details

| Component | Version |
|---|---|
| GPU | Intel Arc Pro B50 Graphics (BMG, PCI ID `8086:e212`) |
| IGC (`libigc2`) | 2.27.10-1222~24.04 |
| `intel-opencl-icd` | 26.01.36711.4-1~24.04~ppa1 |
| `libze-intel-gpu1` | 26.01.36711.4-1~24.04~ppa1 |
| `triton-xpu` | 3.6.0 |
| PyTorch | 2.10.0+xpu |
| Python | 3.12.12 |
| OS | Ubuntu 24.04.4 LTS, kernel 6.17.0-14-generic |

**Comment by etiotto** [comment]
@ofirzaf curious what is the motivation behind setting IGC_ForceOCLSIMDWidth ?

**Comment by whitneywhtsang** [comment]
The `IGC_ForceOCLSIMDWidth=16` option is not understandable by Triton, which causes information mismatch.  
If you want to change threads per warp, you can use the warp_size Triton config.
```
import triton
import triton.language as tl
import torch

@triton.jit
def trivial_masked_store(out_ptr, N: tl.constexpr):
    j = tl.arange(0, N)
    mask = j < 59
    tl.store(out_ptr + j, j, mask=mask)

device = torch.device("xpu")
out = torch.full((64,), -999, dtype=torch.int32, device=device)
trivial_masked_store[(1,)](out, N=64, warp_size=16)
torch.xpu.synchronize()
print(out.cpu().tolist())
```

**Comment by ofirzaf** [comment]
The reason to use `IGC_ForceOCLSIMDWidth=16` was to reduce memory usage. A specific triton kernel used in speculative decoding in vllm required to load tensors in the size of the vocabulary. This kernel caused an OOM error in the scratchpad. So the solution was to use `IGC_ForceOCLSIMDWidth=16`:  vllm-project/vllm#30538

**Comment by whitneywhtsang** [comment]
@ofirzaf can you please confirm if vllm can use Triton config to set warp size like in https://github.com/intel/intel-xpu-backend-for-triton/issues/6206#issuecomment-3977413937?

**Comment by ofirzaf** [comment]
@whitneywhtsang I am not sure. Right now there is no issue in vLLM after removing this environment variable

**Comment by whitneywhtsang** [comment]
@ofirzaf Given that adding Triton config can resolve the provided reproducer, can we close this issue?

**Comment by ofirzaf** [comment]
I think that if the behavior of Triton didn't change when using `IGC_ForceOCLSIMDWidth=16` and you don't intend to fix it, you should at least add a warning when using Triton with this environment variable

**Comment by whitneywhtsang** [comment]
IGC_ForceOCLSIMDWidth is an IGC environment variable, not a Triton environment variable. Triton does not handle or provide warnings for other components' environment variables. This should be used at your own risk and is not intended for production use.
